packages/core/src/indexing-store/cache.ts

```diff
@@ -626,50 +684,120 @@ export const createIndexingCache = ({
         }
       }
     },
-    commit() {
-      for (const [table, tableSpillover] of spillover) {
-        const tableCache = cache.get(table)!;
-        for (const [key, entry] of tableSpillover) {
-          tableCache.set(key, entry);
+    async prefetch({ events, db, eventCount }) {
+      if (isCacheComplete) {
+        if (cacheBytes < common.options.indexingCacheMaxBytes) {
+          return;
         }
-        tableSpillover.clear();
-      }
-
-      cacheBytes += spilloverBytes;
-      spilloverBytes = 0;
 
-      let cacheSize = 0;
-      for (const { size } of cache.values()) cacheSize += size;
+        isCacheComplete = false;
+        for (const table of cache.keys()) {
+          cache.get(table)!.clear();
+          // Note: spillover is not cleared because it is an invariant
+          // it is empty
+        }
+      }
 
-      const flushIndex =
-        totalCacheOps -
-        cacheSize * (1 - common.options.indexingCacheEvictRatio);
+      // Use historical accesses + next event batch to determine which
+      // rows are going to be accessed, and preload them into the cache.
 
-      if (cacheBytes + spilloverBytes > common.options.indexingCacheMaxBytes) {
-        isCacheComplete = false;
+      const prediction = new Map<Table, Map<CacheKey, Row>>();
 
-        let rowCount = 0;
+      for (const table of tables) {
+        prediction.set(table, new Map());
+      }
 
-        for (const tableCache of cache.values()) {
-          for (const [key, entry] of tableCache) {
-            if (entry.operationIndex <= flushIndex) {
-              tableCache.delete(key);
-              cacheBytes -= entry.bytes;
-              rowCount += 1;
+      for (const event of events) {
+        if (profile.has(event.name)) {
+          for (const table of tables) {
+            for (const [key, { count, pattern }] of profile
+              .get(event.name)!
+              .get(table)!) {
+              // Expected value of times the prediction will be used.
+              const ev = count / eventCount[event.name]!;
+              if (ev > 0.25) {
+                prediction
+                  .get(table)!
+                  .set(key, recoverProfilePattern(pattern, event));
+              }
             }
           }
         }
+      }
 
-        common.logger.debug({
-          service: "indexing",
-          msg: `Evicted ${rowCount} rows from the cache`,
-        });
+      for (const [table, tableCache] of cache) {
```

Comment:
So, this block means - if one of our predictions is already in the cache, no need to preload it?

Reply:
yes